tl

```diff
@@ -538,7 +693,10 @@ if cmd == "run" then
    end
 
    tl.loader()
-   return chunk((unpack or table.unpack)(arg))
+   local success,res = pcall(chunk,(unpack or table.unpack)(arg))
+   cleanup_generated_files()
```

Comment:
Now I noticed one thing... is this PR running `build.tl` for _every_ possible `tl` command? I don't think we want to run it for every use of `tl run`. `tl run` is not intended to mean "run this project" as in `cargo run`, but instead it's more like "run this file" as in the Lua interpreter. 

Before looking at this PR, I assumed that only `tl build` would run the build file.

Reply:
check needs to run it as well, as otherwise it can't type check code depending on the generated files.

I can change it so that it doesn't run on the run command though if that is preferred

Reply:
Ermmm, `tl check` and `tl gen` also run file-by-file... it would be annoying to run it multiple times for different files and have them run auxiliary arbitrary scripts multiple times. `tl build` is already the command that means "do everything that needs to be done".

Are the results of `build.tl` preserved after running it once?

If so, Given a `build.tl`, I think a sensible procedure would be to expect users to run `tl build` once on a project, where it would build everything that needs to be pre-built, and then they would be able to run `tl check`, `tl run` and friends without regenerating everything.

In this first PR, I think we could do without a check that `tl build` was actually run first when needed and rely on the user to call things in the right order. As a future improvement, we could have a lock file of sorts in the project dir to detect that `build.tl` was already executed. How does that sound?

Side question: is the writer of `build.tl` expected to write it so it is idempotent?

Reply:
I think the various editor plugins rely on `tl check` though? Which is why I think that `tl check` should execute the build script if one is found.

> Are the results of build.tl preserved after running it once?

Not yet, that was planned to happen in another PR.

> Side question: is the writer of build.tl expected to write it so it is idempotent?

Short answer: Yes.

Longer explanation: Once caching works, the user is able to list a set of patterns where if a file changed that matches these patterns will invalidate the cache and thus trigger the script to rerun. If no files that match these patterns changed since last run (nor the build script itself) then teal will not rerun the script and reuse the previous files.



I can look into making caching happen in this PR, though before I do that I would like to have the point above resolved (The one about using the tmp folder vs a folder in the project).

Reply:
> I would like to have the point above resolved (The one about using the tmp folder vs a folder in the project).

If we want to eventually have caching, then it should be a folder in the project, because folders under /tmp can end up being auto-cleaned-up by system services.

Reply:
> I think the various editor plugins rely on tl check though? 

Yes, but editor plugins which launch compilers doing typecheck-as-you-type don't normally run multi-step build procedures generating files etc, right? For example, if I'm editing a C program in a `cmake` based project, I don't expect it to run the whole build just because I opened a file to edit it.

> If we want to eventually have caching, then it should be a folder in the project, because folders under /tmp can end up being auto-cleaned-up by system services.

Also, I don't mind starting with things in /tmp as they are in this PR but then moving later to the project when caching is added, but I think that at this time `build.tl` should be triggered only by `tl build`.

Reply:
> Yes, but editor plugins which launch compilers doing typecheck-as-you-type don't normally run multi-step build procedures generating files etc, right? For example, if I'm editing a C program in a cmake based project, I don't expect it to run the whole build just because I opened a file to edit it.

If I open a rust project, then yes that is pretty much what happens. `rust-analyzer` runs `cargo check` which results in cargo to download the dependencies if needed and build everything (though it won't spit any actual binaries out as it stops before that phase). This does include running build files however.

> Also, I don't mind starting with things in /tmp as they are in this PR but then moving later to the project when caching is added, but I think that at this time build.tl should be triggered only by tl build.

I already started to move things out of `/tmp` and started work on the caching system. Right now it only checks if he build script changed before running again. I haven't pushed it yet as there are still some bugs left.

Reply:
> If I open a rust project, then yes that is pretty much what happens. rust-analyzer runs cargo check which results in cargo to download the dependencies if needed and build everything (though it won't spit any actual binaries out as it stops before that phase). This does include running build files however.

One thing to note is that `cargo` is a full-fledged package manager. Right now `tl build` is the "odd one out" in the `tl` CLI because it does "package-manager-like things" you'd expect from `cargo` or `npm`, while things like `tl gen`, `tl run`, `tl check` are more analogous to `rustc` or `node`. 

I'm not opposed in principle to having `tl` have a broad scope (`tl build` is already in, after all), but I still want to have the "low-level commands" that operate on the files they are given as arguments. I'm not surprised if `tl build` does a lot of things (and someday even downloads dependencies and whatnot) but I would never expect `tl check file.tl` to do anything other than running the typechecker on `file.tl` and certainly not running some of the project's arbitrary code.

> I think the various editor plugins rely on tl check though? Which is why I think that tl check should execute the build script if one is found.

Most (all?) plugins were written before `tl build` existed. I'm inclined to think that what's needed is to make editor plugins aware of `tl build`, instead of adding build logic into `tl check`.

Reply:
what about being able to pass an argument to teal to also run the build script? And/or have an extra command that runs the build script if needed? In both cases `tl build` would of course still run it if needed as that one is meant to do everything to get the project up and running.

Reply:
> what about being able to pass an argument to teal to also run the build script? And/or have an extra command that runs the build script if needed? In both cases tl build would of course still run it if needed as that one is meant to do everything to get the project up and running.

All of these options are fine by me! I'd just prefer the "low-level commands" that take on a `.tl` file as an argument (`check`, `gen`, `run`) to not do it automatically.

Reply:
decided to go with an argument.